app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt

```diff
@@ -0,0 +1,610 @@
+        WindowManager.LayoutParams().apply {
+            type = WindowManager.LayoutParams.TYPE_APPLICATION_PANEL
+            token = anchor.applicationWindowToken
+            width = WindowManager.LayoutParams.MATCH_PARENT
+            height = WindowManager.LayoutParams.MATCH_PARENT
+            format = PixelFormat.TRANSLUCENT
+            flags = WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN or
+                WindowManager.LayoutParams.FLAG_HARDWARE_ACCELERATED
+        }
+
+    /**
+     * Intended to allow querying the insets of the navigation bar.
+     * Value will be `0` except for when the screen is rotated by 90 degrees.
+     */
+    private fun getLeftInsets() = ViewCompat.getRootWindowInsets(anchor)
+        ?.getInsets(WindowInsetsCompat.Type.systemBars())?.left
+        ?: 0.coerceAtLeast(0)
+
+    @Px
+    private fun Dp.toPx(): Int {
+        return this.value
+            .dpToPx(anchor.resources.displayMetrics)
+            .roundToInt()
+    }
+}
+
+/**
+ * Simpler [View.OnAttachStateChangeListener] only informing about
+ * [View.OnAttachStateChangeListener.onViewDetachedFromWindow].
+ */
+private class OnViewDetachedListener(val onDismiss: () -> Unit) : View.OnAttachStateChangeListener {
+    override fun onViewAttachedToWindow(v: View?) {
+        // no-op
+    }
+
+    override fun onViewDetachedFromWindow(v: View?) {
+        onDismiss()
+    }
+}
+
+/**
+ * Complete content of the popup.
+ * [CFRPopupShape] with a gradient background containing [text] and a dismiss ("X") button.
+ *
+ * @param text String offering more context about a particular feature to the user.
+ * @param indicatorDirection The direction the indicator arrow is pointing to.
+ * @param indicatorArrowStartOffset Maximum distance between the popup start and the indicator arrow.
+ * If there isn't enough space this could automatically be overridden up to 0.
+ * @param indicatorArrowHeight How tall the indicator arrow should be.
+ * @param cornerRadius Corner radius for the popup body.
+ * This will also affect how wide the base of the indicator arrow will be.
+ * @param elevation The z-coordinate at which to place the popup.
+ * This controls the size of the shadow below the popup.
+ * @param onDismiss  Callback for when the popup is dismissed indicating also if the dismissal
+ * was explicit - by tapping the "X" button or not.
+ * @param action Optional other composable to show just below the popup text.
+ */
+@Composable
+@Suppress("LongParameterList", "LongMethod")
+private fun CFRPopupContent(
+    text: String,
+    indicatorDirection: IndicatorDirection,
+    indicatorArrowStartOffset: Dp,
+    indicatorArrowHeight: Dp,
+    cornerRadius: Dp,
+    elevation: Dp = 0.dp,
+    onDismiss: () -> Unit,
+    action: @Composable (() -> Unit)? = null
+) {
+    FirefoxTheme {
+        Box(modifier = Modifier.width(POPUP_WIDTH_DP.dp + HORIZONTAL_PADDING.dp * 2)) {
+            Surface(
```

Comment:
A `Surface` is a wrapper for a `Box`, so the two containers could be merged here by just having the `.width` Modifier before `.align`

Reply:
[`align` is scoped to Box](https://cs.android.com/androidx/platform/tools/dokka-devsite-plugin/+/master:testData/compose/source/androidx/compose/foundation/layout/Box.kt;l=219?q=interface%20BoxScope) and I need that to position the `X` button.
Then I use the Surface because it knows to easily handle a shape plus elevation.

Reply:
Oh dang ok, fair enough. Weird that they didn't trickle down the functionality to `Surface`.